lib/improver/nbhood_percentiles.py

```diff
@@ -0,0 +1,375 @@
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Module containing neighbourhood processing percentiles classes."""
+
+import iris
+from iris.exceptions import CoordinateNotFoundError
+import numpy as np
+
+from improver.utilities.cube_manipulation import concatenate_cubes
+from improver.nbhood import Utilities
+from improver.percentile import PercentileConverter
+
+# Maximum radius of the neighbourhood width in grid cells.
+MAX_RADIUS_IN_GRID_CELLS = 500
+
+
+class NeighbourhoodPercentiles(object):
+    """
+    Apply a neigbourhood processing method to each 2D slice in a cube.
+
+    When applied to a cube, it samples all points in a neighbourhood as
+    equal realizations and derives the requested percentile distribution.
+
+    The neighbourhood methods will presently only work with projections in
+    which the x grid point spacing and y grid point spacing are constant
+    over the entire domain, such as the UK Standard Grid projection
+
+    """
+
+    def __init__(self, method, radii, lead_times=None,
+                 ens_factor=1.0,
+                 percentiles=PercentileConverter.DEFAULT_PERCENTILES):
+        """
+        Create a neighbourhood processing plugin that applies a smoothing
+        to points in a cube.
+
+        Parameters
+        ----------
+
+        method : str
+            Name of the method to use. Options: circular_numpy.
+        radii : float or List (if defining lead times)
+            The radii in metres of the neighbourhood to apply.
+            Rounded up to convert into integer number of grid
+            points east and north, based on the characteristic spacing
+            at the zero indices of the cube projection-x and y coords.
+        lead_times : None or List (optional)
+            List of lead times or forecast periods, at which the radii
+            within 'radii' are defined. The lead times are expected
+            in hours.
+        ens_factor : float (optional)
+            The factor with which to adjust the neighbourhood size
+            for more than one ensemble member.
+            If ens_factor = 1.0 this essentially conserves ensemble
+            members if every grid square is considered to be the
+            equivalent of an ensemble member.
+            Optional, defaults to 1.0
+        percentiles : list (optional)
+            Percentile values at which to calculate; if not provided uses
+            DEFAULT_PERCENTILES from percentile module.
+        """
+        self.percentiles = percentiles
+        self.method_key = method
+        methods = {
+            "circular_numpy": CircularKernelNumpy}
+        try:
```

Comment:
Why is the method called circular_numpy. Something more descriptive might be better e.g. circular_percentiles